QueryBuilder/Dynamic/JoinQuery.cs

```diff
@@ -0,0 +1,82 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+namespace Microsoft.DigitalWorkplace.DigitalTwins.QueryBuilder.Dynamic
+{
+    using System;
+    using System.Collections.Generic;
+    using Microsoft.DigitalWorkplace.DigitalTwins.QueryBuilder.Common.Clauses;
+    using Microsoft.DigitalWorkplace.DigitalTwins.QueryBuilder.Common.Statements;
+
+    /// <summary>
+    /// A wrapper for the common functionalities of query builder.
+    /// </summary>
+    public abstract class JoinQuery<TQuery, TWhereStatement> : FilteredQuery<TQuery, TWhereStatement>
+        where TQuery : JoinQuery<TQuery, TWhereStatement>
+        where TWhereStatement : WhereBaseStatement<TWhereStatement>
+    {
+        private readonly JoinWithStatement<TWhereStatement> joinStatement;
+
+        internal JoinQuery(string rootTwinAlias, IList<string> allowedAliases, SelectClause selectClause, FromClause fromClause, IList<JoinClause> joinClauses, WhereClause whereClause) : base(rootTwinAlias, allowedAliases, selectClause, fromClause, joinClauses, whereClause)
+        {
+            joinStatement = new JoinWithStatement<TWhereStatement>(whereClause, rootTwinAlias);
+        }
+
+        /// <summary>
+        /// Adds a JOIN statement to the query.
+        /// </summary>
+        /// <param name="joinLogic">The functional logic of the JOIN statement containing the required components to join twins via a relationship.</param>
+        /// <returns>The query instance.</returns>
+        public TQuery Join(Func<JoinWithStatement<TWhereStatement>, JoinFinalStatement<TWhereStatement>> joinLogic)
+        {
+            var join = joinLogic.Invoke(joinStatement);
+            return Join(join.Options);
+        }
+
+        /// <summary>
+        /// Adds a JOIN statement to the query with an included set of WHERE clauses.
+        /// </summary>
+        /// <param name="joinAndWhereLogic">The functional logic of the JOIN statement containing the required components to join twins via a relationship.</param>
+        /// <returns>The query instance.</returns>
+        public TQuery Join(Func<JoinWithStatement<TWhereStatement>, WhereCombineStatement<TWhereStatement>> joinAndWhereLogic)
```

Comment:
So, there's 2 parts to my answer for this.
1. I'm not sure if your example would be legal
2. ....because I did not do extensive constraint tests, at least in the way the previously written constraint tests were written wherein they used reflection to confirm if a particular method was available. I'll need to add Where and Join statement specific constraint tests that are unrelated to the larger query objects themselves.